### PR TITLE
Bai 1-Tutorial screen

### PR DESCRIPTION
**Pull request by at-thinhnguyen2**
create tutorial screen

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/ViewModel/Tutorial/TutorialViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+//  Created by Thinh Nguyen X. on 11/23/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import Foundation
+import MVVM
+
+final class TutorialViewModel: MVVM.ViewModel {
+    var photoPaths: [String?] = ["img_1", "img_2", "img_3"]
```
Comment (at-linhvo) [inline MyApp/ViewModel/Tutorial/TutorialViewModel.swift]: add linespace.
rename `var photoNames`

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+import UIKit
+import MVVM
+import SwiftUtils
+
+private struct Config {
+    static let itemWidth = kScreenSize.width
+    static let itemHeight = kScreenSize.height
+}
+final class TutorialViewController: ViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    // MARK: - IBOutlet
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    @IBOutlet weak var skipButton: UIButton!
+    @IBOutlet weak var nextButton: UIButton!
+    @IBOutlet weak var scrollView: UIScrollView!
+    @IBOutlet weak var pageControl: UIPageControl!
+    // MARK - Properties
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    var viewModel = TutorialViewModel()
+    // MARK: - Life Cycle
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configScrollView()
+        configPageControl()
+        setupActions()
+    }
+    func configScrollView() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        configImage()
+        let widthContent = Config.itemWidth * CGFloat(viewModel.photoCount)
+        let heightContent = scrollView.height
+        scrollView.contentSize = CGSize(width: widthContent, height: heightContent)
+        scrollView.isScrollEnabled = viewModel.photoCount > 1
+        scrollView.delegate = self
+        pageControl.numberOfPages = viewModel.photoCount
+        pageControl.addTarget(self, action: Selector(("changePage:")), for: .valueChanged)
+    }
+    func configImage() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            imageView.frame = CGRect(x: x, y: 0, width: Config.itemWidth, height: scrollView.height)
+            imageView.contentMode = .scaleToFill
+            if let nameImage = viewModel.photoPaths[i] {
+                imageView.image = UIImage(named:nameImage)
+            }
+            scrollView.contentSize.width = scrollView.width * CGFloat(i + 1)
+            scrollView.addSubview(imageView)
+        }
+    }
+    func configPageControl() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        pageControl.backgroundColor = .clear
+        pageControl.pageIndicatorTintColor = App.Color.gray184
+        pageControl.currentPageIndicatorTintColor = App.Color.yellow206
+    }
+    func changePage(sender: AnyObject) {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        let x = CGFloat(pageControl.currentPage) * scrollView.width
+        scrollView.setContentOffset(CGPoint(x: x, y :0), animated: true)
+    }
+}
+// MARK: - UIScrollViewDelegate
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+extension TutorialViewController: UIScrollViewDelegate {
+    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
+        let pageNumber = round(scrollView.contentOffset.x / scrollView.width)
+        pageControl.currentPage = Int(pageNumber)
+    }
+}
+// MARK: - Private
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+extension TutorialViewController {
+    fileprivate func setupActions() {
+        skipButton.addTarget(self, action: #selector(skip), for: .touchUpInside)
+        nextButton.addTarget(self, action: #selector(goHome), for: .touchUpInside)
+    }
+    @objc private func skip() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        let homeVC = HomeViewController()
+        AppDelegate.shared.window?.rootViewController = homeVC
+    }
+    @objc private func goHome() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: add linespace

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        let pageWidth = self.scrollView.width
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: remove self

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        let maxWidth = pageWidth * CGFloat(viewModel.photoCount)
+        let contentOffset = self.scrollView.contentOffset.x
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: remove self

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        let slideToX = contentOffset + pageWidth
+        if  slideToX == maxWidth {
+            let homeVC = HomeViewController()
+            AppDelegate.shared.window?.rootViewController = homeVC
+        }
+        self.scrollView.scrollRectToVisible(CGRect(x:slideToX,
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: remove self

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+                                                   width:pageWidth,
+                                                   height:self.scrollView.height),
+                                            animated: true)
+        let pageNumber = round(scrollView.contentOffset.x / scrollView.width)
+        pageControl.currentPage = Int(pageNumber + 1)
+    }
+}
+extension UIScrollView {
+    var height: CGFloat {
+        return self.bounds.size.height
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: remove self

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    }
+    var width: CGFloat {
+        return self.bounds.size.width
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: remove self

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+        self.scrollView.scrollRectToVisible(CGRect(x:slideToX,
+                                                   y:0,
+                                                   width:pageWidth,
+                                                   height:self.scrollView.height),
+                                            animated: true)
+        let pageNumber = round(scrollView.contentOffset.x / scrollView.width)
+        pageControl.currentPage = Int(pageNumber + 1)
+    }
+}
+extension UIScrollView {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: define extension in folder Ext

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+import UIKit
+
+class HomeViewController: UIViewController {
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = App.Color.red214
+        checkTheFistRun()
+    }
+    func checkTheFistRun() {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: add linespace

File: MyApp/View/Controller/Home/HomeViewController.swift [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+//  HomeViewController.swift
+//  MyApp
+//
+//  Created by Thinh Nguyen X. on 11/24/17.
+//  Copyright © 2017 Asian Tech Co., Ltd. All rights reserved.
+//
+
+import UIKit
+
+class HomeViewController: UIViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controller/Home/HomeViewController.swift]: final class

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+import MVVM
+import SwiftUtils
+
+private struct Config {
+    static let itemWidth = kScreenSize.width
+    static let itemHeight = kScreenSize.height
+}
+final class TutorialViewController: ViewController {
+    // MARK: - IBOutlet
+    @IBOutlet weak var skipButton: UIButton!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: `private weak`

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    @IBOutlet weak var nextButton: UIButton!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: `private weak`

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+    @IBOutlet weak var scrollView: UIScrollView!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: `private weak`

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+    @IBOutlet weak var pageControl: UIPageControl!
```
Comment (at-linhvo) [inline MyApp/View/Controller/Tutorial/TutorialViewController.swift]: `private weak`

File: MyApp/View/Base/ViewController.swift [hunk 26]
```diff
@@ -9,4 +9,23 @@
 import UIKit
 import MVVM
 
-class ViewController: UIViewController, MVVM.View { }
+class ViewController: UIViewController, MVVM.View {
```
Comment (at-linhvo) [inline MyApp/View/Base/ViewController.swift]: final class

File: MyApp/View/Base/ViewController.swift [hunk 27]
```diff
@@ -9,4 +9,23 @@
...
+    // MARK - Life cycle
+    init() {
```
Comment (at-linhvo) [inline MyApp/View/Base/ViewController.swift]: no need???

File: MyApp/View/Base/ViewController.swift [hunk 28]
```diff
@@ -9,4 +9,23 @@
...
+        let nibName = String(describing: type(of: self))
+        if let _ = Bundle.main.path(forResource: nibName, ofType: "nib") {
+            super.init(nibName: nibName, bundle: nil)
+        } else {
+            super.init(nibName: nil, bundle: nil)
+        }
+    }
+    override init(nibName: String?, bundle: Bundle?) {
```
Comment (at-linhvo) [inline MyApp/View/Base/ViewController.swift]: no need???

File: MyApp/View/Base/ViewController.swift [hunk 29]
```diff
@@ -9,4 +9,23 @@
...
+        super.init(nibName: nibName, bundle: bundle)
+    }
+    required init?(coder aDecoder: NSCoder) {
```
Comment (at-linhvo) [inline MyApp/View/Base/ViewController.swift]: no need???

File: MyApp/View/Controller/Tutorial/TutorialViewController.swift (hunk 13 above)
Comment (at-thinhnguyen2) [reply MyApp/View/Controller/Tutorial/TutorialViewController.swift]: screen shot 2017-12-05 at 5 01 59 pm
if add linespace SwiftLints Trailing

**Pull request merged**